Implement server-side apply upgrade and downgrade

julianvmodesto opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

##### (1) Allow client-side to server-side apply upgrade.

Ensure that a user can change management of an object from client-side apply to
server-side apply without conflicts.

We do this by reading the last-applied-configuration annotation for client-side apply
to identify if conflicts can be forced because `kubectl apply` owns them.

##### (2) Allow server-side apply to client-side downgrade.

For an object managed with client-side apply, a user may upgrade to
managing the object with server-side apply, then decide to downgrade.

We can support this downgrade by keeping the last-applied-configuration
annotation for client-side apply updated with server-side apply.

This was suggested by @apelisse. This is an alternative to conflicting entirely and failing a downgrade.

See this KEP update: https://github.com/kubernetes/enhancements/pull/1382

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related to https://github.com/kubernetes/kubernetes/issues/80916, https://github.com/kubernetes/kubernetes/issues/90146

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Support a smooth upgrade from client-side apply to server-side apply without conflicts, as well as support the corresponding downgrade.
```

julianvmodesto committed:
Implement server-side apply upgrade & downgrade.

- Allow client-side to server-side apply upgrade.

  Ensure that a user can change management of an object from client-side apply to
  server-side apply without conflicts.

- Allow server-side apply to client-side downgrade.

  For an object managed with client-side apply, a user may upgrade to
  managing the object with server-side apply, then decide to downgrade.

  We can support this downgrade by keeping the last-applied-configuration
  annotation for client-side apply updated with server-side apply.

julianvmodesto commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,80 @@
+	delete(annotations, corev1.LastAppliedConfigAnnotation)
+	accessor.SetAnnotations(annotations)
+
+	newLastApplied, err := json.Marshal(obj)
```
Oh wait this doesn't work, I think I need runtime.Encode here

fedebongio commented:
/assign @kwiesmueller

kwiesmueller commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go:
```diff
@@ -0,0 +1,102 @@
+		t.Fatalf("last applied annotation is empty")
+	}
+
+	expected := `
```
I think that newline is failing your test.

julianvmodesto replied:
Hm I'm going to move this test somewhere else since the test field manager doesn't use the `merge`

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,85 @@
+	}
+
+	delete(annotations, corev1.LastAppliedConfigAnnotation)
+	accessor.SetAnnotations(annotations)
```
I don't think these two lines are needed or even good here. You should probably just update rather than delete and re-insert.

julianvmodesto replied:
We need to remove the last-applied annotation when we serialize the object since it will be in the annotation itself. Here's the corresponding client-side apply method:

https://github.com/kubernetes/kubernetes/blob/875d3b534e3f6f118c3bf7f7e5a6b4c6ac6fe92a/staging/src/k8s.io/kubectl/pkg/util/apply.go#L71-L74

apelisse replied:
Ah yes, your right! Thanks :-)

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,85 @@
+	newLastApplied, err := runtime.Encode(unstructured.UnstructuredJSONScheme, obj)
+	if err != nil {
+		annotations[corev1.LastAppliedConfigAnnotation] = oldLastApplied
+		accessor.SetAnnotations(annotations)
```
Ah, you re-inject here in case of an error, then why did you delete it? :-)

apelisse reviewed (commented):
> This is an alternative to conflicting entirely and failing a downgrade.

I don't know if that's what this does. This updates the last applied annotation so that if you want to re-use CSA after SSA, you can. BUT, I don't think any of this will avoid the conflicts with CSA on your first SSA call?

I think it's a good step, but I don't think that's enough, and we need to be careful how we do it.

@lavalamp, do you have an opinion on maintaining the last-applied annotation so that someone can re-use CSA?

julianvmodesto commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,85 @@
+// Apply keeps the client-side apply annotation up to date with
+// server-side apply managed fields
+func (f *lastAppliedUpdater) Apply(liveObj runtime.Object, newObj runtime.Object, managed Managed, manager string, force bool) (runtime.Object, Managed, error) {
+	err := f.updateLastApplied(newObj)
```
Oops, I think we need to serialize `newObj` as the annotation, but we need to set the annotation on `liveObj` to make this work

apelisse replied:
Correct!

julianvmodesto replied:
Okay, fixed. Will need an e2e or integration test for this.

Tested on a local cluster:

```
$ cat <<EOF > test.yaml
→ apiVersion: v1
→ kind: Pod
→ metadata:
→   namespace: default
→   name: pause
→ spec:
→   containers:
→   - image: k8s.gcr.io/pause:3.1
→     name: pause
→ EOF

$ ./cluster/kubectl.sh apply -f test.yaml
pod/pause created

$ ./cluster/kubectl.sh apply view-last-applied -f test.yaml
apiVersion: v1
kind: Pod
metadata:
  annotations: {}
  name: pause
  namespace: default
spec:
  containers:
  - image: k8s.gcr.io/pause:3.1
    name: pause

$ sed s/pause:3.1/pause:latest/g test.yaml -i

$ ./cluster/kubectl.sh apply -f test.yaml --server-side --force-conflicts
pod/pause serverside-applied

$ ./cluster/kubectl.sh apply view-last-applied -f test.yaml
apiVersion: v1
kind: Pod
metadata:
  name: pause
  namespace: default
spec:
  containers:
  - image: k8s.gcr.io/pause:latest
    name: pause

```

julianvmodesto commented:
> > This is an alternative to conflicting entirely and failing a downgrade.
> 
> I don't know if that's what this does. This updates the last applied annotation so that if you want to re-use CSA after SSA, you can. BUT, I don't think any of this will avoid the conflicts with CSA on your first SSA call?

Oh no, I was referencing the downgrade, not the upgrade. For downgrade from SSA to CSA, we could make CSA fail for an object that's being managed by SSA [that @jennybuckley described in the downgrade proposal](https://docs.google.com/document/d/1_YKsMttYZmIw2cSg-jH5PExHjA7FUyALAgHuXDDBcFY/edit#heading=h.2u66gomiwpuj) https://github.com/kubernetes/kubernetes/issues/90146: 

> CSA requests need to be rejected by the server if the object being applied to is already in the new format. This is slightly more difficult, since CSA requests look just like a StrategicMergePatch or a MergePatch (depending on what is supported). We can get around this by checking what apply format the object is in, in each of those patch types, and then failing if the patch is attempting to set the “lastApplied” annotation in an SSA format object.

This change you proposed will make it so that SSA to CSA downgrade is valid by keeping the last-applied annotation up to date.

apelisse commented:
OK, Right, yes I agree then! :-)

lavalamp commented:
Why do we want to keep the client side annotation? How does that help the downgrade? What does the downgrade look like if we don't maintain this?

lavalamp commented:
I agree we should support downgrade, but we should probably support it for all resources, and we definitely don't want to track the client side annotation for all resources.

I guess I expected downgrade to be something like:
* client sends old apply type (patch with annotation set) ... and maybe a --force-conflicts? We don't want people to do this accidentally
* server deletes managed fields

lavalamp commented:
If we wanted to get super fancy, we could notice a GET from a kubectl client side apply command, and fabricate a client side annotation at that time. I don't recommend this but it's possible.

apelisse commented:
> I agree we should support downgrade, but we should probably support it for all resources

What do you mean "all resources"? All objects in etcd or all resource types?
If you mean "all objects in etcd", I don't think it makes sense to create something new for objects that have never been applied. I think the goal is only to allow people who used CSA in the past to keep using it (even though they maybe gave a try, possibly by accident, to SSA).

The goal here is mostly:
1. People have used CSA to apply a specific object, it has the last-applied annotation,
2. They try to apply that object with SSA and it works
3. They try to apply again but using CSA (maybe because they have different users applying, maybe because they applied with SSA by accident). I think this should work. As long as we've kept the last-applied annotation in sync with what CSA would have done, it will work, as long as there is only one "applier" and one version. 

> server deletes managed fields

Not sure which managed fields? All of them? That'd be losing all the "history", which means they wouldn't be able to nicely go back to SSA.

> If we wanted to get super fancy, we could notice a GET from a kubectl client side apply command, and fabricate a client side annotation at that time. I don't recommend this but it's possible.

I don't think we would even know which manager to use to build the annotation from? And also we don't really have the values anymore.

lavalamp commented:
> They try to apply again but using CSA (maybe because they have different users applying, maybe because they applied with SSA by accident). I think this should work. 

Why? I think it should be possible, but I don't think it should be transparent. Users should have to explicitly downgrade. The point of this whole exercise is to get them onto SSA... Users should only be going backwards if there's a bug in SSA they need to work around, IMO.

apelisse commented:
Yeah, I think they will organically as they update their version of kubectl, since it's going to be the default. But I don't think all the users are going to have the same version, so it's very possible that they go back-and-forth without even realizing. I don't see any reason to not have the smoothest experience in that case.

lavalamp commented:
> But I don't think all the users are going to have the same version, so it's very possible that they go back-and-forth without even realizing.

I don't think we're helping people if we make it work that way. People should know what their system is doing. Wouldn't you want to know if your coworker upgraded to a SSA version before they were supposed to?

Also people should definitely not be using the existing client side apply like that, it's not really safe for multiple appliers.

apelisse commented:
> Also people should definitely not be using the existing client side apply like that, it's not really safe for multiple appliers.

Maybe it's a bad example, but whoever had time to re-deploy the mungebot would do it :-)
I wouldn't be surprised if it's still a super common workflow.

apelisse commented:
> Wouldn't you want to know if your coworker upgraded to a SSA version before they were supposed to?

Not that much, I think it's going to happen by accident anyway. I'm a little more concerned about a coworker using CSA and maybe stomping some field they shouldn't. But even that I think is better than the alternatives I remember seeing.

lavalamp commented:
> whoever had time to re-deploy the mungebot would do it :-)

...and you'd definitely want to know if the last person used SSA, don't you think?

apelisse commented:
> ...and you'd definitely want to know if the last person used SSA, don't you think?

I guess the question is, what difference does it make?

lavalamp commented:
It adds risk to do something different. Teams should generally be using tools at the same versions, there's just weird problems if you don't. It makes onboarding coworkers harder when they get random, different error messages.

apelisse commented:
OK So, disregarding how it's done, what sort of experience do you think would be better?

I can think of a couple:
1. We error when going from CSA to SSA and require some sort of "force". That's sort-of what we're doing today, though today you'll get an error when you change fields owned by CSA, which means you can get a conflict two years after starting to use SSA, when you finally change that field that was still owned by CSA.
2. We error when going from SSA to CSA and you can force, but we can't provide that flag through older versions of kubectl. I think what makes the most sense is: if I force, I can then apply repeatedly with CSA without forcing, and then if I re-SSA (smoothly), I would need to force again to go back to CSA mode.

lavalamp commented:
Yeah, the latter sounds close to what I'd expect--

Ideally, transitioning from CSA -> SSA would be seamless, but if it leaves
a landmine like you mention, then it'd be much less confusing to get a
force up-front.

Transitioning from SSA -> CSA should require a force, since we hope you
only want to do that in an exceptional case.

In either case, if you continue to use the tool that last did the apply,
you shouldn't have to force (unless someone else clobbered your object, of
course).

On Fri, Apr 17, 2020 at 4:24 PM Antoine Pelisse <notifications@github.com>
wrote:

> OK So, disregarding how it's done, what sort of experience do you think
> would be better?
>
> I can think of a couple:
>
>    1. We error when going from CSA to SSA and require some sort of
>    "force". That's sort-of what we're doing today, though today you'll get an
>    error when you change fields owned by CSA, which means you can get a
>    conflict two years after starting to use SSA, when you finally change that
>    field that was still owned by CSA.
>    2. We error when going from SSA to CSA and you can force, but we can't
>    provide that flag through older versions of kubectl. I think what makes the
>    most sense is: if I force, I can then apply repeatedly with CSA without
>    forcing, and then if I re-SSA (smoothly), I would need to force again to go
>    back to CSA mode.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/90187#issuecomment-615505491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAE6BFQG6AFUYWYLW5S3J23RNDQJFANCNFSM4MJBIYPA>
> .
>

kwiesmueller commented:
> We error when going from SSA to CSA and you can force, but we can't provide that flag through older versions of kubectl.

Why can't we use the already existing force flag in kubectl for this?

Also linking my previous comment from the other discussion: https://github.com/kubernetes/kubernetes/issues/90146#issuecomment-613692487

kwiesmueller commented:
And I agree that people should be using the same tools/versions etc.
I'm not sure if that's going on out there though. I'm quite sure many people are mixed with my teams. And as there is no warning or anything when your kubectl doesn't match the server or the ones previously writing the object, i wouldn't expect a lot of users to care :-/

apelisse commented:
>> We error when going from SSA to CSA and you can force, but we can't provide that flag through older versions of kubectl.
>
> Why can't we use the already existing force flag in kubectl for this?

The force flag in kubectl doesn't alter the request or does anything we want. I would be strongly opposed to changing the semantics of the flag by cherry-picking changes to older versions.

kwiesmueller commented:
Ah, i didn't know it's not touching the request.
Then i agree.

julianvmodesto commented:
Rebased in https://github.com/kubernetes/kubernetes/pull/84134 and converted the ApplyWithLastApplied (upgrader) to a Manager because we must look at last-applied (for the upgrade) before changing last-applied as in this change (for the downgrade).

apelisse commented:
One of your test is failing for a good reason.

julianvmodesto commented:
Fixed integration test + also added unit tests

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,102 @@
+		return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
+	}
+	setLastApplied(newObj)
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
```
That could be simpler?
```golang
{
	if manager == "kubectl" {
	    setLastApplied(newObj)
	}
	return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
 }
```

julianvmodesto replied:
Oh nice, done

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,101 @@
+func (f *lastAppliedUpdater) Apply(liveObj, newObj runtime.Object, managed Managed, manager string, userAgent string, force bool) (runtime.Object, Managed, error) {
+	// Sync the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	if manager != "kubectl" {
```
it should be `==`, not `!=`, right? You want to set the last applied annotation if it *is* kubectl?

apelisse replied:
Make sure your test failed by the way, if it didn't then I'm concerned :D

julianvmodesto replied:
Gahh yes

apelisse replied:
That test failed, it's probably related?
https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/90187/pull-kubernetes-bazel-test/1273731844889645060

julianvmodesto replied:
Fixed! Confirmed the tests failed as expected

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,101 @@
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
+}
+
+func setLastApplied(obj runtime.Object) {
```
Why did you decide not to return an error here?

apelisse replied:
Since this is in "Apply", we have more tolerance to failure.

julianvmodesto replied:
Ah okay, returning an error here. Not sure if we should add any further error remediation message like you mentioned for the upgrade? e.g. ask user to remove the last-applied annotation or set it properly with kubectl subcommand

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:
```diff
@@ -0,0 +1,159 @@
+	if err != nil {
+		return newLiveObj, newManaged, newErr
+	}
+	if !conflictSet.Difference(allowedConflictSet).Empty() {
```
We probably discussed that already, but I forgot the outcome. Should we really complain about things that we believe shouldn't be conflict? In other words, shouldn't we return the difference as the set of conflicts here?

julianvmodesto replied:
Oh wow right, we should do this and only return the difference

julianvmodesto replied:
Done

julianvmodesto replied:
Added a test for this

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:
```diff
@@ -0,0 +1,159 @@
+	return f.fieldManager.Apply(liveObj, newObj, managed, manager, "", true)
+}
+
+func (f *lastAppliedManager) allowedConflictsFromLastApplied(liveObj runtime.Object) (*fieldpath.Set, error) {
```
I was wondering what solution would exist for people who have a non-parseable last-applied annotation. There already exists tool in kubectl to edit it, but I also want to mention that forcing apply should solve the problem. Would you be able to confirm that it would not result in this piece of code failing (I think it wouldn't because we only get here if there are conflicts) and that the last-applied annotation would be updated none-the less?

julianvmodesto replied:
Added a test for this, `TestServerSideApplyWithInvalidLastApplied`

apelisse reviewed (commented):
Thanks again, that's pretty awesome!

apelisse commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,104 @@
+func (f *lastAppliedUpdater) Apply(liveObj, newObj runtime.Object, managed Managed, manager string, userAgent string, force bool) (runtime.Object, Managed, error) {
+	// Sync the client-side apply annotation
+	// only if the manager is the default manager for kubectl server-side apply
+	if manager == ServerSideApplyFieldManager {
```
Why are we using the manager rather than the userAgent here?

julianvmodesto replied:
Oh, fixed and commented as discussed.

apelisse commented:
/retest
/lgtm
/approve

apelisse commented:
Looks like the errors are spurious and/or timing-out.

apelisse reviewed (commented):
/lgtm

julianvmodesto commented:
This change implements both the upgrade and downgrade. Consolidated from https://github.com/kubernetes/kubernetes/pull/84134.

/assign @lavalamp
/retest

apelisse commented:
/approve
/lgtm

apelisse commented:
/assign @lavalamp

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedmanager.go:
```diff
@@ -0,0 +1,173 @@
+	newLiveObj, newManaged, newErr := f.fieldManager.Apply(liveObj, newObj, managed, manager, userAgent, force)
+	// Upgrade the client-side apply annotation only from kubectl server-side-apply.
+	// We use the user agent because the field manager from kubectl may be customized.
+	if userAgent != "kubectl" {
```
If the user agent is customized, will the user really expect to do the upgrade logic? We definitely don't want more than one applier triggering the upgrade logic (right?), how is that avoided?

julianvmodesto replied:
Replaced with a check for the server-side apply field manager instead of the kubectl user agent, as discussed with @kubernetes/release-managers 

Users may opt-out by specifying a different field manager, which we will document.

lavalamp commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater_test.go:
```diff
@@ -0,0 +1,91 @@
+		t.Errorf("failed to get last applied: %v", err)
+	}
+
+	if lastApplied == originalLastApplied ||
```
I'd feel more comfortable if this were checked golden data, or then actually run through the kubectl CSA logic to ensure that it's actually compatible.

julianvmodesto replied:
The integration test in test/cmd usees kubectl directly -- is this okay? Maybe we can do something better in test/integration/apiserver/apply/apply_test.go too.

lavalamp replied:
Yeah that's probably good enough. As long as we would notice if it broke.

lavalamp commented:
Sorry for delay in reviewing... Apart from what I commented on, looks great :)

lavalamp commented:
/approve

I think that made the change cleaner, too, thanks!

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
Oh I’ll probably need to squash

apelisse commented:
Awesome @julianvmodesto !
/lgtm

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

apelisse commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

julianvmodesto commented:
/retest

apelisse commented:
/retest

julianvmodesto commented:
/retest

apelisse commented:
I'm confused, is `/retest` retrying things that have succeeded in the past? Every time you run `/retest` you have new tests failing? This is never going to end, will it? :-)

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/fieldmanager/lastappliedupdater.go:
```diff
@@ -0,0 +1,117 @@
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to build last-applied annotation: %v", err)
+		}
+		err = setLastApplied(liveObj, lastAppliedValue)
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to set last-applied annotation: %v", err)
+		}
```
this needs to check if the new annotation value causes https://github.com/kubernetes/kubectl/issues/712 and if so, I think it should clear the annotation rather than populating annotations that fail validation

julianvmodesto replied:
Ah will send a change
